xmr-btc/src/alice.rs

```diff
@@ -1,31 +1,167 @@
+use crate::{
+    bitcoin,
+    bitcoin::{BroadcastSignedTransaction, WatchForRawTransaction},
+    bob, monero,
+    monero::{CreateWalletForOutput, Transfer},
+    transport::{Receive, Send},
+};
 use anyhow::{anyhow, Result};
-use ecdsa_fun::adaptor::{Adaptor, EncryptedSignature};
+use ecdsa_fun::{
+    adaptor::{Adaptor, EncryptedSignature},
+    nonce::Deterministic,
+};
 use rand::{CryptoRng, RngCore};
-
-use crate::{bitcoin, bitcoin::GetRawTransaction, bob, monero, monero::ImportOutput};
-use ecdsa_fun::{nonce::Deterministic, Signature};
 use sha2::Sha256;
+use std::convert::{TryFrom, TryInto};
+
+pub mod message;
+pub use message::{Message, Message0, Message1, Message2, UnexpectedMessage};
+
+pub async fn next_state<
+    R: RngCore + CryptoRng,
+    B: WatchForRawTransaction + BroadcastSignedTransaction,
+    M: CreateWalletForOutput + Transfer,
+    T: Send<Message> + Receive<bob::Message>,
+>(
+    bitcoin_wallet: &B,
+    monero_wallet: &M,
+    transport: &mut T,
+    state: State,
+    rng: &mut R,
+) -> Result<State> {
+    match state {
+        State::State0(state0) => {
+            transport
+                .send_message(state0.next_message(rng).into())
+                .await?;
 
-#[derive(Debug)]
-pub struct Message0 {
-    pub(crate) A: bitcoin::PublicKey,
-    pub(crate) S_a_monero: monero::PublicKey,
-    pub(crate) S_a_bitcoin: bitcoin::PublicKey,
-    pub(crate) dleq_proof_s_a: cross_curve_dleq::Proof,
-    pub(crate) v_a: monero::PrivateViewKey,
-    pub(crate) redeem_address: bitcoin::Address,
-    pub(crate) punish_address: bitcoin::Address,
+            let bob_message0: bob::Message0 = transport.receive_message().await?.try_into()?;
+            let state1 = state0.receive(bob_message0)?;
+            Ok(state1.into())
+        }
+        State::State1(state1) => {
+            let bob_message1: bob::Message1 = transport.receive_message().await?.try_into()?;
+            let state2 = state1.receive(bob_message1);
```

Comment:
I believe `state1.receive` only expects one specific type: `bob::Message1` right? Shouldn't need the type annotation then :)